### ISSUE TITLE
support package level annotations

### ISSUE THREAD

**Issue by rabbah** [nimbella/nimbella-cli#7]
This is feedback from a chat with @akrabat in Slack.

The deployer supports package (and project) level environment variables and parameters but annotations are per action. I think this has some historically roots (parameters are merged between package and action but annotations are not).

This issue is to consider adding annotations at package level, at least.

**Comment by joshuaauerbachwatson** [comment]
We have annotations at package level:   https://nimbella.io/downloads/nim/nim.html#annotations-1

I'm confused about what is being requested here.   Do you mean to make package level annotations distribute over actions, the way parameters do?   That would violate the current OW information model.   Please clarify.

**Comment by rabbah** [comment]
> you mean to make package level annotations distribute over action

I believe this was the intent, yes. But let me refine further - we currently permit a `web` annotation at the package level. This itself is syntactic sugar for an annotation. Based on the slack exchange, I think what was desired was actually another web action related modifier: `raw-http`.

If we did add annotations at package level that distribute over actions, I don't think we need to modify the backend implementation. It could be done client side, in the deployer, no?

**Comment by akrabat** [comment]
From my point of view, having to add:

```
        annotations:
          raw-http: true
```

to every single action in the project.yml is tedious and makes for a harder to read file than adding it once at the `package` level.

**Comment by joshuaauerbachwatson** [comment]
>  we currently permit a `web` annotation at the package level. This itself is syntactic sugar for an annotation. 

What we permit is a `web` _property_ at package level in `package.yml`.   Yes, informally it is an "annotation" but it isn't an annotation in the OW sense.   It is not syntactic sugar for a single annotation.   It behaves like the `--web` flag in the `wsk` CLI which means it expands into useful combinations of the `web-export`, `final`, and `raw-http` annotations.

```
    if (action.web == true) {
        annotations['web-export'] = true
        annotations['final'] = true
        annotations['raw-http'] = false
    } else if (action.web === "raw") {
        annotations['web-export'] = true
        annotations['final'] = true
        annotations['raw-http'] = true
    } else if (action.web == false) {
        annotations['web-export'] = false
        annotations['final'] = false
        annotations['raw-http'] = false
    }
```

> Based on the slack exchange, I think what was desired was actually another web action related modifier: `raw-http`.

But we already have that (see above).  Just specify
```
web: raw
```

at package level and it will distribute over the actions as designed.

> If we did add annotations at package level that distribute over actions, I don't think we need to modify the backend implementation. It could be done client side, in the deployer, no?

Yes, I agree.   In suggesting that this _general_ capability would violate the OW information model, I meant that we already have package level annotations which attach to the package itself (and don't distribute to the actions).   These have a use.  So, we would violate the OW model by changing the meaning `annotations:` at package level.   If we wanted a _general_ capability, to distribute annotations, we'd need to distinguish these from regular old package annotations (syntactically).

> having to add:
> 
> ```
>         annotations:
>           raw-http: true
> ```
> 
> to every single action in the `project.yml` is tedious and makes for a harder to read file than adding it once at the `package` level.

Agreed.  You can get that effect today by putting `web: raw` at package level.

In summary, I believe the _narrow_ issue as expressed by Rob requires no action.   I am not sure what the motivation would be to generalize this capability to all annotations.   There are relatively few annotations that are both user-specifiable and actually affect the semantics of the actions (the three covered by `web` are among them).

**Comment by akrabat** [comment]
Thanks. Will test.
